plugins/base/src/main/kotlin/transformers/documentables/PropertiesMergerTransformer.kt

```diff
@@ -0,0 +1,86 @@
+            val (functions, properties) = mergePotentialBeansAndField(this.functions, this.properties)
+            this.copy(functions = functions, properties = properties)
+        }
+        is DInterface -> {
+            val (functions, properties) = mergePotentialBeansAndField(this.functions, this.properties)
+            this.copy(functions = functions, properties = properties)
+        }
+        is DObject -> {
+            val (functions, properties) = mergePotentialBeansAndField(this.functions, this.properties)
+            this.copy(functions = functions, properties = properties)
+        }
+        is DAnnotation -> {
+            val (functions, properties) = mergePotentialBeansAndField(this.functions, this.properties)
+            this.copy(functions = functions, properties = properties)
+        }
+        is DPackage -> {
+            val (functions, properties) = mergePotentialBeansAndField(this.functions, this.properties)
+            this.copy(functions = functions, properties = properties)
+        }
+        else -> this
+    } as T
+
+    private fun DFunction.getPropertyNameForFunction() =
+        when {
+            JvmAbi.isGetterName(name) -> propertyNameByGetMethodName(Name.identifier(name))?.asString()
+            JvmAbi.isSetterName(name) -> propertyNamesBySetMethodName(Name.identifier(name)).firstOrNull()
+                ?.asString()
+            else -> null
+        }
+
+    private fun mergePotentialBeansAndField(
+        functions: List<DFunction>,
+        fields: List<DProperty>
+    ): Pair<List<DFunction>, List<DProperty>> {
+        val fieldNames = fields.associateBy { it.name }
+        val accessors = mutableMapOf<DProperty, MutableList<DFunction>>()
+        val regularMethods = mutableListOf<DFunction>()
+        functions.forEach { method ->
+            val field = method.getPropertyNameForFunction()?.let { name -> fieldNames[name] }
+            if (field != null) {
+                accessors.getOrPut(field, ::mutableListOf).add(method)
+            } else {
+                regularMethods.add(method)
+            }
+        }
+        return regularMethods.toList() to accessors.map { (dProperty, dFunctions) ->
+            if (dProperty.visibility.values.all { it is KotlinVisibility.Private }) {
+                dFunctions.flatMap { it.visibility.values }.toSet().singleOrNull()?.takeIf {
+                    it in listOf(KotlinVisibility.Public, KotlinVisibility.Protected)
+                }?.let { visibility ->
+                    dProperty.copy(
+                        getter = dFunctions.firstOrNull { it.type == dProperty.type },
+                        setter = dFunctions.firstOrNull { it.parameters.isNotEmpty() },
+                        visibility = dProperty.visibility.mapValues { visibility }
+                    )
+                } ?: dProperty
+            } else {
+                dProperty
+            }
+        } + fields.toSet().minus(accessors.keys.toSet())
```

Comment:
This is more or less copied from PsiToDocumentableTranslator. I wonder if we could remove it from there since this transformer should work well for PSIs and Descriptor sourced Documentables. However, I don't have time to experiment with that :(

Reply:
Lines 581 - 603: functions `getPropertyNameForFunction` and `splitFunctionsAndAccessors`

These are losely copied and tailored to our needs here, yet it was copy-pasted. I will annotate it with todo as you asked